sequtils: simplify `toSeq` implementation

zerbina opened the pull request:
## Summary

Don't use complex macros, but instead use `compiles` + generic routines
for achieving the same. This should slightly improve compile times for
code often using `toSeq`, and is also another step towards less macro
usage during bootstrapping.

## Details

This change partially reverts https://github.com/nim-works/nimskull/pull/151,
which replaced the usage of `compiles` with macros. However, compared to
macros, `compiles` is the less problematic feature usage here, allowing
the implementation to be significantly less complex. In addition, the
`compiles` here is also easier to replace with type-inspection routines
in the future (such as `x is iterator`, once that works).

zerbina commented on lib/pure/collections/sequtils.nim:
```diff
@@ -844,10 +784,16 @@ macro toSeq*(arg: untyped): untyped =
+  when compiles(typeof(items(arg))):
+    # arg is something that supports the ``items`` iterator:
+    toSeqBuiltin[typeof(items(arg))](arg)
+  elif compiles(typeof(arg())):
```
This should be turned into `arg is iterator`, once the compiler support testing for iterator types that.

saem reviewed (approved):
lesson learned... hopefully. 🤣 

Type driven is better than AST driven... especially as you said once we have the ability to test for `iterator`.

zerbina committed:
sequtils: simplify `toSeq` implementation

Summary
=======

Don't use complex macros, but instead use `compiles` + generic routines
for achieving the same. This should slightly improve compile times for
code often using `toSeq`, and is also another step towards less macro
usage during bootstrapping.

Details
=======

This change partially reverts https://github.com/nim-works/nimskull/pull/151,
which replaced the usage of `compiles` with macros. However, compared to
macros, `compiles` is the less problematic feature usage here, allowing
the implementation to be significantly less complex. In addition, the
`compiles` here is also easier to replace with type-inspection routines
in the future (such as `x is iterator`, once that works).

saem commented:
/merge

saem commented:
/merge

saem commented:
/merge

looks like a spurious failure
